Adding gtests for Color and Matrix

SgStapleton opened the pull request.

SgStapleton committed:
Adding gtests for Color and Matrix

SgStapleton committed:
Updating gtest for Matrix

SgStapleton committed:
More work to ColorTests.cpp and MatrixTests.cpp

SgStapleton committed:
Added test for constructor making use of TNT:Array2D in MatrixTests

SgStapleton committed:
Merging upstream testing

jessemapel commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+  TEST (MatrixTests, ConstructorWithDimensions) {
+    Matrix A(2, 2, 2);
+    
+    ASSERT_EQ(2, A[0][0]);
```
These should be EXPECT_EQ so that all failures get checked. You should also assert that the matrix has the proper dimensions.
```suggestion
    ASSERT_EQ(2, A.Rows());
    ASSERT_EQ(2, A.Columns());
    EXPECT_EQ(2, A[0][0]);
```

jessemapel commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+    Matrix A(2, 2, 2);
+    
+    ASSERT_EQ(2, A[0][0]);
+    ASSERT_EQ(2, A[0][1]);
```
```suggestion
    EXPECT_EQ(2, A[0][1]);
```

jessemapel commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+    
+    ASSERT_EQ(2, A[0][0]);
+    ASSERT_EQ(2, A[0][1]);
+    ASSERT_EQ(2, A[1][0]);
```
```suggestion
    EXPECT_EQ(2, A[1][0]);
```

jessemapel commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+    ASSERT_EQ(2, A[0][0]);
+    ASSERT_EQ(2, A[0][1]);
+    ASSERT_EQ(2, A[1][0]);
+    ASSERT_EQ(2, A[1][1]);
```
```suggestion
    EXPECT_EQ(2, A[1][1]);
```

jessemapel commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Invalid matrix dimensions"));
```
No need to do toLatin1() http://doc.qt.io/qt-5/qstring.html#contains

krlberry replied:
Could you update this to something like:

```  
EXPECT_TRUE(e.toString().toLatin1().contains("Invalid matrix dimensions")) << e.toString().toStdString() ;
```
? 

This makes the error easier to understand if the test does fail. (Without this, it just will say:   Actual: false
Expected: true)

krlberry commented on isis/tests/ColorTests.cpp:
```diff
@@ -0,0 +1,81 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Can not convert an invalid color to an RGBA string.  There is no string representation of an invalid color"));
```
Could you update this to something like:

```  
EXPECT_TRUE(e.toString().toLatin1().contains("Can not convert an invalid color to an RGBA string.  There is no string representation of an invalid color")) << e.toString().toStdString() ;
```
? 

This makes the error easier to understand if the test does fail. (Without this, it just will say:   Actual: false
Expected: true)

krlberry commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Invalid matrix dimensions"));
```
Could you update this to something like:

```  
EXPECT_TRUE(e.toString().toLatin1().contains("Invalid matrix dimensions")) << e.toString().toStdString()  ;
```
? 

This makes the error easier to understand if the test does fail. (Without this, it just will say:   Actual: false
Expected: true)

krlberry commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Invalid matrix dimensions"));
```
Could you update this to something like:

```  
EXPECT_TRUE(e.toString().toLatin1().contains("Invalid matrix dimensions")) << e.toString().toStdString()  ;
```
? 

This makes the error easier to understand if the test does fail. (Without this, it just will say:   Actual: false
Expected: true)

krlberry commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Invalid matrix dimensions"));
```
Could you update this to something like:

```  
EXPECT_TRUE(e.toString().toLatin1().contains("Invalid matrix dimensions")) << e.toString().toStdString()  ;
```
? 

This makes the error easier to understand if the test does fail. (Without this, it just will say:   Actual: false
Expected: true)

krlberry commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Unable to calculate the determinant, the matrix is not square"));
```
Could you update this to something like:

```  
EXPECT_TRUE(e.toString().toLatin1().contains("Invalid matrix dimensions")) << e.toString().toStdString()  ;
```
? 

This makes the error easier to understand if the test does fail. (Without this, it just will say:   Actual: false
Expected: true)

krlberry commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Unable to calculate the trace, the matrix is not square"));
```
Could you update this to something like:

```  
EXPECT_TRUE(e.toString().toLatin1().contains("Invalid matrix dimensions")) << e.toString().toStdString()  ;
```
? 

This makes the error easier to understand if the test does fail. (Without this, it just will say:   Actual: false
Expected: true)

krlberry commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Incompatible matrix dimensions, cannot add the matrices"));
```
Same comment about updating to stream the expected value to `EXPECT_TRUE`.

krlberry commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Incompatible matrix dimensions, cannot add the matrices"));
```
Same comment about updating to stream the expected value to `EXPECT_TRUE`.

krlberry commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Incompatible matrix dimensions, cannot subtract the matrices"));
```
Same comment about updating to stream the expected value to `EXPECT_TRUE`.

krlberry commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Incompatible matrix dimensions, cannot subtract the matrices"));
```
Same comment about updating to stream the expected value to `EXPECT_TRUE`.

krlberry commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Incompatible matrix dimensions, cannot multiply the matrices"));
```
Same comment about updating to stream the expected value to `EXPECT_TRUE`.

krlberry commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Incompatible matrix dimensions, cannot multiply the matrices"));
```
Same comment about updating to stream the actual exception error message to `EXPECT_TRUE`.

krlberry commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Incompatible matrix dimensions, cannot multiply the matrices"));
```
Same comment about updating to stream the actual value to `EXPECT_TRUE`.

krlberry commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Unable to calculate the inverse, the matrix is not square"));
```
Same comment about updating to stream the actual value to `EXPECT_TRUE`.

krlberry commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Unable to calculate eigenvectors, the matrix is not square"));
```
Same comment about updating to stream the actual value to `EXPECT_TRUE`.

krlberry commented on isis/tests/MatrixTests.cpp:
```diff
@@ -0,0 +1,381 @@
+      FAIL() << "Expected an IException";
+    }
+    catch(IException &e) {
+      EXPECT_TRUE(e.toString().toLatin1().contains("Unable to calculate eigenvalues, the matrix is not square"));
```
Same comment about updating to stream the actual value to `EXPECT_TRUE`.

kaitlyndlee commented on isis/tests/ColorTests.cpp:
```diff
@@ -0,0 +1,81 @@
+  }
+  
+  
+  TEST (ColorTests, InvalidColorToString) {
```
Do we want spaces between TEST and () to follow ISIS standards?

krlberry replied:
@kdl222 This is a good question. If we do decide to do this, we should update the in-progress [Getting Started with testing](https://github.com/USGS-Astrogeology/ISIS3/wiki/Getting-started-with-testing-(gtest)) guide.

jessemapel replied:
Doesn't the ISIS style guide say no space? https://isis.astrogeology.usgs.gov/documents/CodingStandards/CodingStandards.html#methodDefinitions

jessemapel committed:
Update isis/tests/MatrixTests.cpp

Co-Authored-By: SgStapleton <sgstapleton@usgs.gov>

jessemapel committed:
Update isis/tests/MatrixTests.cpp

Co-Authored-By: SgStapleton <sgstapleton@usgs.gov>

jessemapel commented:
I could see a case for creating a custom predicate assertion and formatter for testing exception messages. See https://github.com/abseil/googletest/blob/master/googletest/docs/advanced.md#predicate-assertions-for-better-error-messages and https://github.com/abseil/googletest/blob/master/googletest/docs/advanced.md#using-a-predicate-formatter

It would reduce duplication and make it easier for us to keep all of our exception testing consistent.

jessemapel commented:
Possible predicate formatter:

```
::testing::AssertionResult AssertIExceptionContains(
      const char* e_expr,
      const char* contents_expr,
      IException &e,
      QString contents) {
  if (e.toString().contains(contents)) return ::testing::AssertionSuccess();

  return ::testing::AssertionFailure() << e_expr << "\'s error message (" << e.toString()
      << ") does not contain " << contents_expr << " (" << contents << ").";
}
```

This is then used with

```
EXPECT_PRED_FORMAT2(AssertIExceptionContains, e, "Substring to look for in error message");
```

krlberry commented:
@jessemapel :+1: I really like this idea. One comment about the specific code suggested -- does `e.toString()` work? I had to change this to `e.toStdString()` in some earlier code.

SgStapleton commented:
@kberryUSGS @jessemapel  I will implement and test this out!

jessemapel commented:
@kberryUSGS It probably does need to use toStdString()

krlberry commented:
@jessemapel @SgStapleton I have an additional thought about this. I haven't tested this yet, but the codes we return with the error message do mean something and if we're making a custom predicate formatter anyway, we could do something like the following: 

```
EXPECT_PRED_FORMAT2(AssertIExceptionContains, e, "Substring to look for in error message", 2);
```

```
::testing::AssertionResult AssertIExceptionContains(
      const char* e_expr,
      const char* contents_expr,
      IException &e,
      QString contents
      int errorCode ) {
  if (e.toString().contains(contents) && e.errorType() ==errorCode) return ::testing::AssertionSuccess();
  
if (!e.toString().contains(contents) {
  return ::testing::AssertionFailure() << e_expr << "\'s error message (" << e.toString()
      << ") does not contain " << contents_expr << " (" << contents << ").";
} 

if(e.errorType()!=errorCode) {
return ::testing::AssertionFailure() << e_expr << "\'s error code << QString::number(e.errorType()) << 
       " does not equal " << QString::number(errorCode); 
}
```

Thoughts?

jessemapel commented:
@kberryUSGS I think that's fine. I would move the check into a separate predicate function because it's getting rather complex.

```
bool checkIException(IException, &e, QString contents, int errorCode) {
  return e.toString().contains(contents) && e.errorType() == errorCode
}
```

SgStapleton committed:
Making recommended changes

SgStapleton committed:
Merging Jesse's suggestions

SgStapleton committed:
Fixing odd GitHub behavior...

There were suggested changes made in PR that were not playing nicely. Could not push correct version for whatever reason.
